### PR TITLE
MM-15324 Allow plugins override ephemeral posts

### PR DESCRIPTION
**Pull request by scottleedavis**
#### Summary
This PR enables ephemeral posts with a post.prop.type to override a post by a webapp plugin.

e.g. keeps the system_ephemeral post.type, but allows post.props.type to override
```
		p.API.SendEphemeralPost(args.UserId, &model.Post{
			UserId:    p.botId,
			ChannelId: args.ChannelId,
			Message:   "Bot ephemeral link",
			Props: model.StringInterface{
				"type": "system_ephemeral_test_plugin",
			},
		})
```

This has been validated by using `/test-ephemeral-post-override` from the [test bot plugin](https://github.com/scottleedavis/mattermost-plugin-bot-test/releases/tag/v0.1.4)

#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-server/issues/10797

#### Notes

I did not add a test to post_message_view_test.jsx as I wasn't sure if it was needed and I am unfamiliar with snapshot testing.  (happy to do it if requested).
I did not add a test for post_body.jsx as no tests currently exist for the component.

### CODE REVIEW

File: components/post_view/post_body/post_body.jsx [hunk 1]
```diff
@@ -173,7 +173,7 @@ export default class PostBody extends React.PureComponent {
             </React.Fragment>
         );
 
-        const hasPlugin = post.type && this.props.pluginPostTypes.hasOwnProperty(post.type);
+        const hasPlugin = (post.type && this.props.pluginPostTypes.hasOwnProperty(post.type)) || (post.props && post.props.type);
```
Comment (saturninoabril) [inline components/post_view/post_body/post_body.jsx]: Should the `post.props.type` be checked with `pluginPostTypes.hasOwnProperty`?

Comment (scottleedavis) [reply components/post_view/post_body/post_body.jsx]: I think so.  Good idea.

**Comment by hanzei** [comment]
@cpoile Gently reminder about this PR. Would be great if we can get this in v5.12.

**Comment by scottleedavis** [comment]
Thanks @cpoile !  That is correct.  I chose this route over change the post type to have a prefix 'system_ephemeral_<custom_name>', as a lot of things would need to change, and I wanted to preserve the behavior with as little change as possible.

**Comment by jasonblais** [comment]
@hanzei @scottleedavis does this need any documentation at https://developers.mattermost.com/extend/plugins/?

**Comment by scottleedavis** [comment]
@jasonblais I think it would be a good idea.  Otherwise it isn't straight forward to use props to solve the issue.  I haven't done documentation updates yet, happy to do so if requested.  :)

**Comment by jasonblais** [comment]
@scottleedavis It would be great if you had time to help with the doc! :)

**Comment by aaronrothschild** [comment]
Documented via: https://github.com/mattermost/mattermost-webapp/pull/3024 . Thanks Scott!

**Pull request merged**